src/System.Linq/tests/ConcatTests.cs

```diff
@@ -89,142 +70,138 @@ public void SecondNull()
             Assert.Throws<ArgumentNullException>("second", () => Enumerable.Range(0, 0).Concat(null));
         }
 
-        [Fact]
-        public void TwoEnumerableSources()
+        [Theory]
+        [MemberData(nameof(ArraySourcesData))]
+        [MemberData(nameof(ArraySelectSourcesData))]
+        [MemberData(nameof(EnumerableSourcesData))]
+        [MemberData(nameof(NonCollectionSourcesData))]
+        [MemberData(nameof(ListSourcesData))]
+        [MemberData(nameof(ConcatOfConcatsData))]
+        [MemberData(nameof(ConcatWithSelfData))]
+        [MemberData(nameof(ChainedCollectionConcatData))]
+        public void VerifyEquals(IEnumerable<int> expected, IEnumerable<int> actual)
         {
-            VerifyEquals(
-                Enumerable.Range(0, 7),
-                Enumerable.Range(0, 3).Concat(Enumerable.Range(3, 4)));
+            // workaround: xUnit type inference doesn't work if the input type is not T (like IEnumerable<T>)
+            VerifyEqualsWorker(expected, actual);
         }
 
-        [Fact]
-        public void TwoArraySources()
+        private static void VerifyEqualsWorker<T>(IEnumerable<T> expected, IEnumerable<T> actual)
         {
-            VerifyEquals(
-                Enumerable.Range(0, 7),
-                Enumerable.Range(0, 3).ToArray().Concat(Enumerable.Range(3, 4).ToArray()));
+            // All of these transforms should take an enumerable and produce
+            // another enumerable with the same contents.
+            var transforms = new List<EnumerableTransform<T>>
+            {
+                e => e,
+                e => e.ToArray(),
+                e => e.ToList(),
+                e => e.Select(i => i),
+                e => e.Select(i => i).ToArray(),
+                e => e.Where(i => true),
+                e => e.Concat(Array.Empty<T>()),
+                e => ForceNotCollection(e),
+                e => e.Concat(ForceNotCollection(Array.Empty<T>()))
+            };
```

Comment:
I only kept them since we were testing `Where(i => true)` and `Select(i => i)` before this change, and I didn't want to accidentally remove code coverage. I think we can get rid of the `Where` since that's handled by `ForceNotCollection`, and `Select(i => i).ToArray()` should be the same thing as `ToArray()`. I'm keeping the `Select` part though, since that's the only `IIListProvider` that is not a List.